examples/ensemble/plot_gradient_boosting_quantile.py

```diff
@@ -3,77 +3,268 @@
+# target variable. To make this even more interesting we consider
+# heteroschedastic noise where the parameter sigma depends on the input x.
+#
+# The lognormal distribution is very skewed, meaning that it is likely to get
+# large outliers but impossible to observe small outliers.
+sigma = 0.5 + X.ravel() / 10
+noise = rng.lognormal(sigma=sigma) - np.exp(sigma ** 2 / 2)
 y += noise
-y = y.astype(np.float32)
 
-# Mesh the input space for evaluations of the real function, the prediction and
-# its MSE
+# %%
+# Split into train, test datasets:
+X_train, X_test, y_train, y_test = train_test_split(X, y)
+
+# %%
+# Fitting non-linear quantile and least squares regressors
+# --------------------------------------------------------
+#
+# Fit gradient boosting models trained with the quantile loss and
+# alpha=0.05, 0.5, 0.95.
+#
+# The models obtained for alpha=0.05 and alpha=0.95 produce a 90% confidence
+# interval (95% - 5% = 90%).
+#
+# The model trained with alpha=0.5 produces a regression of the median: on
+# average, there should be the same number of target observations above and
+# below the predicted values.
+from sklearn.ensemble import GradientBoostingRegressor
+from sklearn.metrics import pinball_loss, mean_squared_error
+
+
+all_models = {}
+common_params = dict(
+    learning_rate=0.05,
+    n_estimators=250,
+    max_depth=2,
+    min_samples_leaf=9,
+    min_samples_split=9,
+)
+for alpha in [0.05, 0.5, 0.95]:
+    gbr = GradientBoostingRegressor(loss='quantile', alpha=alpha,
+                                    **common_params)
+    all_models["q %1.2f" % alpha] = gbr.fit(X_train, y_train)
+
+# %%
+# For the sake of comparison, also fit a baseline model trained with the usual
+# least squares loss (ls), also known as the mean squared error (MSE).
+gbr_ls = GradientBoostingRegressor(loss='ls', **common_params)
+all_models["ls"] = gbr_ls.fit(X_train, y_train)
+
+# %%
+# Create an evenly spaced set of input values spanning the [0, 10] range.
 xx = np.atleast_2d(np.linspace(0, 10, 1000)).T
 xx = xx.astype(np.float32)
 
-alpha = 0.95
+# %%
+# Plot the true function (expected mean), the prediction of the conditional
+# mean (least squares loss), the conditional median and the conditional 90%
+# interval (from 5th to 95th conditional percentiles).
+import matplotlib.pyplot as plt
 
-clf = GradientBoostingRegressor(loss='quantile', alpha=alpha,
-                                n_estimators=250, max_depth=3,
-                                learning_rate=.1, min_samples_leaf=9,
-                                min_samples_split=9)
 
-clf.fit(X, y)
+y_pred = all_models['ls'].predict(xx)
+y_lower = all_models['q 0.05'].predict(xx)
+y_upper = all_models['q 0.95'].predict(xx)
+y_med = all_models['q 0.50'].predict(xx)
 
-# Make the prediction on the meshed x-axis
-y_upper = clf.predict(xx)
+fig = plt.figure(figsize=(10, 10))
+plt.plot(xx, f(xx), 'g:', label=r'$f(x) = x\,\sin(x)$')
```

Comment:
Can we make the true mean curve, dashed green, a bit thicker?

Reply:
done